proposals/VK_EXT_present_timing.adoc

```diff
@@ -0,0 +1,554 @@
+
+Similarly, when using `presentAtAbsoluteTime` feature to schedule presents at specific times, a present stage must be specified as a target. See <<scheduling>>.
+
+* `VK_PRESENT_STAGE_QUEUE_OPERATIONS_END_BIT_EXT` marks the end of the set of queue operations enqueued by `vkQueuePresentKHR` on the provided `VkQueue`. These queue operations are implementation-specific; the usual example is a blit to a system-specific internal surface suited for presentation.
+* `VK_PRESENT_STAGE_IMAGE_LATCHED_BIT_EXT` is the step after which the image associated with the presentation request has been latched by the presentation engine to create the presentation of a future refresh cycle. For example, in a flip-model scenario, this is the time the presentation request's image has been selected for the next refresh cycle.
+* `VK_PRESENT_STAGE_IMAGE_FIRST_PIXEL_OUT_BIT_EXT` is the stage after which data for the first pixel of the presentation request associated with the image has left the presentation engine for a display hardware.
+* `VK_PRESENT_STAGE_IMAGE_FIRST_PIXEL_VISIBLE_BIT_EXT` is the stage after which a display hardware has made the first pixel visible for the presentation request associated with the image to be presented.
+
+Implementations are required to support at least `VK_PRESENT_STAGE_QUEUE_OPERATIONS_END_BIT_EXT` in `VkSurfacePresentTimingCapabilitiesEXT::presentStageQueries` if `presentTimingSupported` is `VK_TRUE` for the surface.
+
+=== Enabling present timing for a swapchain
+
+To enable present timing for a swapchain, a new flag must be specified in `VkSwapchainCreateInfoKHR::flags`: `VK_SWAPCHAIN_CREATE_PRESENT_TIMING_BIT_EXT`.
+
+To provide presentation timing results, implementations need to allocate an internal queue and other resources to collect the necessary timestamps. The size of that queue must be specified by the application with a new function:
+
+[source,c]
+----
+VkResult vkSetSwapchainPresentTimingQueueSizeEXT(
+    VkDevice                                    device,
+    VkSwapchainKHR                              swapchain,
+    uint32_t                                    size);
+----
+
+Calling this function multiple times causes the results queue to be reallocated to the new size. If the new size cannot hold all the current outstanding results, `VK_NOT_READY` is returned.
+
+Calling `vkQueuePresentKHR` with non-zero stage queries allocates a slot in that internal queue, while `vkGetPastPresentationTimingEXT` releases slots when complete results are returned.
+
+=== Swapchain Timing Information
+
+==== Timing Properties
+
+For timing to be meaningful, the application needs to be aware of various properties. Basic properties are exposed in a new structure, `VkSwapchainTimingPropertiesEXT`, which can be retrieved with:
+
+[source,c]
+----
+VkResult vkGetSwapchainTimingPropertiesEXT(
```

Comment:
It would be nice if there was a "poll" mechanism for this.

That is, an application can register a listener, that the PE can fire as soon as timing properties change.

Reply:
mistyped -- s/poll/push

Reply:
Agreed, and here's my thinking on that: There are a handful of situations in Vulkan where such a mechanism would be nice. Most of them fall into the realm of system integration stuff. I've previously written some extensions that use a VkFence as a sort of notification mechanism, and it works OK from an implementation point of view, but it doesn't integrate well with the spec overall, and I regret it. Callbacks are difficult to get right because they can make locking and API re-entrance complicated. Ultimately, my conclusion was we should have a new object that encapsulates notification of things that aren't necessarily tied to completion of queue work, but I think that's outside the scope of this API. It could easily be added as a supplemental feature in a future extension though.